urlaa051fbd3d_WheezePuppet_collegeDiversity_tgz-pJ8-edu_umw_cpsc_collegediversity_SimJ8/src/edu/umw/cpsc/collegediversity/StatsPrinter.java

```diff
@@ -0,0 +1,66 @@
+package edu.umw.cpsc.collegediversity;
+
+import sim.engine.*;
+import sim.util.Bag;
+import java.io.PrintWriter;
+import java.io.File;
+import java.io.FileWriter;
+
+/**
+ * A utility class to dump periodic statistical output about the
+ * simulation.
+ */
+public class StatsPrinter implements Steppable {
+
+    private static StatsPrinter theInstance;
+
+    /**
+     * Singleton pattern.
+     */
+    public static synchronized StatsPrinter instance() {
+        if (theInstance == null) {
+            theInstance = new StatsPrinter();
+        }
+        return theInstance;
+    }
+
+    private StatsPrinter() {
+    }
+
+    /**
+     * Dump statistical output about the simulation at the end of every
+     * academic year, immediately before students promote/dropout/graduate.
+     * This will run every April 30th. In particular:
+     * <ul>
+     * <li>Create an output file with filename of the form
+     * <code>year2023.csv</code>, where the number is one greater than the
+     * current value returned from {@link Sim#getYear()}. This line will
+     * contain a header line, then one line per active student in the 
+     * academic year ending in that number. (For instance, the file just 
+     * named would include all students enrolled during the 2022-2023 
+     * academic year.) Each line will be of the following form:
+     *     <p><code>243,white,2,239</code><p/>
+     * which says "The student with ID 243, who was a white sophomore, had
+     * 239 connections with other students." (See 
+     * {@link Student#toString}).)</li>
+     * </ul>
+     */
+    public void step(SimState state) {
+        try{
+            PrintWriter out = new PrintWriter(new FileWriter(
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code contains a resource that might not be closed properly. This can cause a resource leak that slows down or crashes your system.

**Fix**
Consider closing the following resource: *out*. The resource is referenced at lines 50, 53, 57. The resource is closed at line 59. There are other execution paths that do not contain closure statements, for example, when Bag.get throws an exception. Ensure *out* is closed in a try-finally block or declared in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).